Adjust which playbook is run depending on which group was changed

pypingou opened the pull request.

pypingou committed:
Adjust which playbook is run depending on which group was changed

pypingou commented:
This hasn't had much testing :(

ralphbean commented:
I don't understand.  What is the difference between the two playbooks?

Also, will this fail on some of the message types?  Do any of them have an odd ``group`` field?  See https://apps.fedoraproject.org/datagrepper/raw?topic=org.fedoraproject.prod.fas.role.update&topic=org.fedoraproject.prod.fas.group.member.sponsor&topic=org.fedoraproject.prod.fas.group.member.remove&topic=org.fedoraproject.prod.fas.user.update for a query.

pypingou commented:
> I don't understand. What is the difference between the two playbooks?

``run_fasClient.yml`` runs on all our hosts but some that do not have fasClient, while ``run_fasClient_simple.yml`` only runs on some of the hosts (fedorapeople, fedorahosted, bastion, pkgs)

ralphbean commented on fedmsg_fasclient.py:
```diff
@@ -89,8 +89,18 @@ def delayed_consume():
+        groups = [
+            msg['msg']['group']
+            for msg in messages
+            if 'sysadmin-' in msg['msg'].get('group', None)
```
Perhaps just ``'sysadmin'`` instead of ``'sysadmin-'``?

pypingou replied:
I was wondering if there are groups having ``sysadmin`` in their name but not related to us (I'm thinking a ``foosysadmingit`` or so.

ralphbean commented:
So this has us doing simple by default.  But the more complete one when the change is with respect to a group.

This deserves an inline comment to explain.  Why is it more complex when there's a group involved?  What kind of change happened that deserves only a simple run when there is no group involved.

Is it that the simple runs take place only when a public SSH key changes?

What about if a sysadmin with shell access changes their SSH key?  That info won't get propagated to all the nodes.

pypingou commented:
- fas.group.member.sponsor has a group
- fas.user.update has no group
- fas.group.member.remove has a group
- fas.role.update has a group

So fas.user.update is the only one giving us problem :-/

pypingou commented:
> What about if a sysadmin with shell access changes their SSH key? That info won't get propagated to all the nodes.

Looking into the data for my previous reply I was thinking about this and maybe we should look into the message for ``fas.user.update`` and use the global playbook if the user updated their ssh key.

Thoughts?

pypingou commented:
>  Why is it more complex when there's a group involved? What kind of change happened that deserves only a simple run when there is no group involved.

It's not about having groups involved, it's about which groups are involved :)

pypingou committed:
Document what's going on and run the whole playbook upon ssh_key changes

pypingou committed:
Reformat doc

pypingou committed:
React only on FAS changes in prod

ralphbean commented on fedmsg_fasclient.py:
```diff
@@ -89,8 +89,32 @@ def delayed_consume():
+        for msg in messages:
+            # If one of the groups includes a sysadmin-related group, run the
+            # complete playbook
+            if 'sysadmin-' in msg['msg'].get('group', None) \
```
Maybe leave the ``'-'`` off the end of ``'sysadmin-'`` so we can catch the "sysadmin" group itself?

pypingou committed:
Search sysadmin instead of sysadmin- and None by []

- Let's be more generous about which group we find
- Don't forget that None is not iterable so use an empty string instead

ralphbean commented:
Nice!  :+1:
